### PR TITLE
[testing] rename skip targets + docs

### LINKED ISSUE #6349: [testing] USE_CUDA default and intuitive skip decorators
This library's primarily use is for gpu work, and currently many tests won't run even if gpu is available, since the current setup wants env var `USE_CUDA` to be true for anything to happen.  It's easy to forget to manually add this env var to pytest command line.

To maximize the testing potential I propose that only if `USE_CUDA=False` then the test is skipped (for CI jobs that need to test library's work on cpu), otherwise if `if torch.cuda.is_available()` cuda tests can be run.

In a brief discussion @julien-c suggested that:

> The original thinking was that we wanted to make sure that when we wanted to run on GPU it actually ran on GPU. i.e. it should even fail if you do `USE_CUDA`  and there's no GPU, to prevent silent failures on GPU

and the discussion stopped there. This ticket was opened to complete this discussion.

@julien-c, could you please share a specific scenario based on the design intention you shared?

also `CUDA_VISIBLE_DEVICES=""` could be used to easily emulate a non-gpu environment if need be, w/o introducing new env vars. i.e. it'd be a built-in equivalent of `USE_CUDA=False`.

Further, `USE_CUDA` is currently only used for skip decorators. This setting cannot currently be respected from within a test. e.g. in a test I'm currently working on I have:

```
        if torch.cuda.is_available():
            testargs += ['--fp16', '--gpus=1']
```
so it'll ignore  `USE_CUDA`, as the test must always run whether there is a gpu or not, so no skip decorator was used. This ignoring conceptually won't do the right thing then as it'll run the test on gpu even if `USE_CUDA==False` (or unset). So if `USE_CUDA`-functionalty remains, there is a need for an accessor that is not a [skip decorator](https://github.com/huggingface/transformers/blob/master/src/transformers/testing_utils.py#L126).

`require_multigpu` also currently ignores `USE_CUDA`

### PR DESCRIPTION
**Pull request by stas00**
As discussed in https://github.com/huggingface/transformers/issues/6349 this PR brings consistency to skip decorators, so now we will have:

* `require_torch` - this test will run only under torch
* `require_torch_gpu` - as `require_torch` plus requires at least 1 GPU
* `require_torch_multigpu` - as `require_torch` plus requires at least 2 GPUs
* `require_torch_non_multigpu` - as `require_torch` plus requires 0 or 1 GPUs
* `require_torch_tpu` - as `require_torch` plus requires at least 1 TPU

Documentation updated and expanded.

The main change was done by running:
```
find . -type d -name ".git" -prune -o -type f -exec perl -pi -e 's|require_multigpu|require_torch_multigpu|g' {} \;
find . -type d -name ".git" -prune -o -type f -exec perl -pi -e 's|require_torch_and_cuda|require_torch_gpu|g' {} \;
find . -type d -name ".git" -prune -o -type f -exec perl -pi -e 's|require_non_multigpu|require_torch_non_multigpu|g' {} \;
```

Fixes: #6349

@LysandreJik, @sgugger, @sshleifer

### CODE REVIEW

**Review by sshleifer — APPROVED** [review]

File: docs/source/testing.rst [hunk 1]
```diff
@@ -393,36 +393,53 @@ On a GPU-enabled setup, to test in CPU-only mode add ``CUDA_VISIBLE_DEVICES=""``
...
+* ``require_torch_multigpu`` - as ``require_torch`` plus requires at least 2 GPUs
+* ``require_torch_non_multigpu`` - as ``require_torch`` plus requires 0 or 1 GPUs
+* ``require_torch_tpu`` - as ``require_torch`` plus requires at least 1 TPU
+
+For example, here is a test that must be run only when there are 2 or more GPUs available and pytorch is installed.
+
+.. code-block:: python
+
+    @require_torch_multigpu
+    def test_example_bar():
```
Comment (sgugger) [inline docs/source/testing.rst:417]:
```suggestion
    def test_example_with_multigpu():
```
nit: foo and bar can be a bit gatekeeping as new coders may not know those terms.

File: docs/source/testing.rst [hunk 2]
```diff
@@ -393,36 +393,53 @@ On a GPU-enabled setup, to test in CPU-only mode add ``CUDA_VISIBLE_DEVICES=""``
...
+
+If a test requires ``tensorflow`` use the ``require_tf`` decorator. For example:
+
+.. code-block:: python
+
+    @require_tf
+    def test_tf_foo():
```
Comment (sgugger) [inline docs/source/testing.rst:424]:
```suggestion
    def test_tf_thing_with_tensorflow():
```

File: docs/source/testing.rst [hunk 3]
```diff
@@ -393,36 +393,53 @@ On a GPU-enabled setup, to test in CPU-only mode add ``CUDA_VISIBLE_DEVICES=""``
...
+
+These decorators can be stacked. For example, if a test is slow and requires at least one GPU under pytorch.
+
+.. code-block:: python
+
+    @require_torch_gpu
+    @slow
+    def test_example_foo():
```
Comment (sgugger) [inline docs/source/testing.rst:432]:
```suggestion
    def test_example_slow_on_gpu():
```

**Review by sgugger — APPROVED** [review]
Thanks a lot for doing this! Just added some nits.

**Comment by stas00** [comment]
I resolved conflicts - it's good to go now.

**Review by LysandreJik — APPROVED** [review]
Very welcome change! Thanks @stas00!

**Pull request merged**